zcash_client_sqlite/src/wallet.rs

```diff
@@ -157,29 +163,92 @@ pub(crate) fn add_account<P: consensus::Parameters>(
     params: &P,
     account: AccountId,
     key: &UnifiedFullViewingKey,
+    birthday: Option<AccountBirthday>,
 ) -> Result<(), SqliteClientError> {
-    add_account_internal(conn, params, "accounts", account, key)
-}
+    // Set the wallet birthday, falling back to the chain tip if not specified
+    let chain_tip = scan_queue_extrema(conn)?.map(|(h, _)| h);
+
+    let birthday_height = birthday.as_ref().map_or_else(
+        || chain_tip.ok_or(SqliteClientError::ChainHeightUnknown),
+        |birthday| Ok(birthday.height()),
+    )?;
 
-pub(crate) fn add_account_internal<P: consensus::Parameters, E: From<rusqlite::Error>>(
-    conn: &rusqlite::Transaction,
-    network: &P,
-    accounts_table: &'static str,
-    account: AccountId,
-    key: &UnifiedFullViewingKey,
-) -> Result<(), E> {
-    let ufvk_str: String = key.encode(network);
     conn.execute(
-        &format!(
-            "INSERT INTO {} (account, ufvk) VALUES (:account, :ufvk)",
-            accounts_table
-        ),
-        named_params![":account": &<u32>::from(account), ":ufvk": &ufvk_str],
+        "INSERT INTO accounts (account, ufvk, birthday_height)
+        VALUES (:account, :ufvk, :birthday_height)",
+        named_params![
+            ":account": u32::from(account),
+            ":ufvk": &key.encode(params),
+            ":birthday_height": u32::from(birthday_height)
+        ],
     )?;
 
+    // If a birthday frontier is available, insert it into the note commitment tree. If the
+    // birthday frontier is the empty frontier, we don't need to do anything.
+    if let Some(frontier) = birthday.as_ref().and_then(|b| b.sapling_frontier().value()) {
+        debug!("Inserting frontier into ShardTree: {:?}", frontier);
+        let shard_store = SqliteShardStore::<
+            _,
+            zcash_primitives::sapling::Node,
+            SAPLING_SHARD_HEIGHT,
+        >::from_connection(conn, SAPLING_TABLES_PREFIX)?;
+        let mut shard_tree: ShardTree<
+            _,
+            { zcash_primitives::sapling::NOTE_COMMITMENT_TREE_DEPTH },
+            SAPLING_SHARD_HEIGHT,
+        > = ShardTree::new(shard_store, PRUNING_DEPTH.try_into().unwrap());
+        shard_tree.insert_frontier_nodes(
+            frontier.clone(),
+            Retention::Checkpoint {
+                id: birthday_height,
+                is_marked: false,
+            },
+        )?;
+    }
+
+    let sapling_activation_height = params
+        .activation_height(NetworkUpgrade::Sapling)
+        .expect("Sapling activation height must be available.");
+
+    // Add the ignored range up to the birthday height. We always re-scan the birthday block, even
+    // though it should never contain funds for the wallet, just to avoid off-by-one errors that
+    // might occur if someone were to use the block height of their first received transaction as
+    // the birthday.
```

Comment:
I think that this is fine, given that we're going to be suggesting that the birthday height be set to `chain_tip - PRUNING_DEPTH` to avoid potential errors related to reorg.